### ISSUE TITLE
withData() not working as expected with BehaviorSpec

### ISSUE THREAD

**Issue by nsk90** [kotest/kotest#3148]
kotest version: 5.4.1
kotlin version: 1.7.10

I have a trouble with data tests using `BehaviorSpec`. I have tried both `withData()` overloads for top level scope and container scope usage.

* top level case fails with InvalidDslException
* container case does not compile util I use explicit context receiver, which looks ugly

```
class BehaviourTest: BehaviorSpec ({

    // this test fails with InvalidDslException: Cannot add a root test after the spec has been instantiated
    withData(nameFn = {"$it"}, true , false) {
        Given("given") {
            When("when") {
                Then("then") {
                    true shouldBe true
                }
            }
        }
    }

    Given("given1") {
        withData(nameFn = {"$it"}, true , false) {
            // this does not compile until using explicit this@Given.
            this@Given.When("when") {
                Then("then") {
                    true shouldBe true
                }
            }
        }
    }
})
```

Does `BehaviorSpec` support using `withData()` api? Is that a bug or I use it in incorrect way?

**Comment by CheshenkoVladislav** [comment]
i have same problem :(

**Comment by AlexSkvor** [comment]
Faced this too, hirts

**Comment by sksamuel** [comment]
`withData` is meant to be your test scope. You don't need to add further nests inside it.
It's not really meant for use with something as structured as given when then.
In that case, just put a for loop instead of the with data, and use the values from the loop in your test names.

**Comment by nsk90** [comment]
Ok, thank you for reply. But what about second overload for usage inside a scope?

**Comment by sksamuel** [comment]
WithData tests are always final state tests.
Until we move away from junit platform this limitation will remain in place unfortunately.

**Comment by sksamuel** [comment]
This will be fixed in 5.5 :)
